File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            map.put(players[i], i);
        }

        for (String call : callings) {
            int idx = map.get(call);
            int targetIdx = idx - 1;

            String temp = players[idx];
            players[idx] = players[targetIdx];
            players[targetIdx] = temp;

            map.put(call, targetIdx);
            map.put(players[idx], idx);
        }

        return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    let result = players;
    let obj = {};
    players.forEach((e, i) => {
        obj[e] = i;
    });
    
    callings.forEach(e => {
        const idx = obj[e];
        const target = idx - 1;
        
        obj[e]--;
        obj[result[target]]++;
        
        const temp = result[idx];
        result[idx] = result[target];
        result[target] = temp;
    }) 
    
    return result;
}